File: ios-native/RAVEN/RAVEN/Core/Protocol/DiscoveryResolver.swift
//
//  DiscoveryResolver.swift
//  RAVEN — multi-lane discovery without central Raven DB (Discovery V1).
//
//  Finding a name ≠ verifying a person. @alias is NOT identity; rvn1… is.
//  Spec: docs/RAVEN_DISCOVERY_V1.md. FastAPI must not be used when serverless ON.
//

import Foundation

enum DiscoveryScope: String, Codable {
    case local = "LOCAL"
    case exactId = "EXACT_ID"
    case exactAlias = "EXACT_ALIAS"
    case myNetwork = "MY_NETWORK"
    case nearby = "NEARBY"
    case publicScope = "PUBLIC"
    case all = "ALL"
}

enum VerificationState: String, Codable, Equatable {
    case directlyVerified = "DIRECTLY_VERIFIED"
    case trustedContact = "TRUSTED_CONTACT"
    case introduced = "INTRODUCED"
    case scopedVerified = "SCOPED_VERIFIED"
    case nearbyVerified = "NEARBY_VERIFIED"
    case publicSignedProfile = "PUBLIC_SIGNED_PROFILE"
    case aliasConflict = "ALIAS_CONFLICT"
    case expiredOrStale = "EXPIRED_OR_STALE"
    case blocked = "BLOCKED"
}

enum DiscoverySource: String, Codable, Hashable {
    case localContacts = "LOCAL_CONTACTS"
    case exactRavenId = "EXACT_RAVEN_ID"
    case aliasDht = "ALIAS_DHT"
    case nearbyBle = "NEARBY_BLE"
    case socialIntroduction = "SOCIAL_INTRODUCTION"
    case publicProfileIndex = "PUBLIC_PROFILE_INDEX"
    case legacyServer = "LEGACY_SERVER"
}

struct DiscoveryIntroduction: Codable, Equatable {
    var introducerRavenId: String
    var subjectRavenId: String
}

struct DiscoveryResult: Codable, Equatable {
    var ravenId: String
    var displayName: String
    var aliases: [String]
    var profileDigest: String
    var sourceSet: [DiscoverySource]
    var verificationState: VerificationState
    var introductions: [DiscoveryIntroduction]
    var conflictCount: UInt32
    var sequence: UInt64
    var expiresAt: UInt64

    enum CodingKeys: String, CodingKey {
        case ravenId = "raven_id"
        case displayName = "display_name"
        case aliases
        case profileDigest = "profile_digest"
        case sourceSet = "source_set"
        case verificationState = "verification_state"
        case introductions
        case conflictCount = "conflict_count"
        case sequence
        case expiresAt = "expires_at"
    }

    /// Stable schema keys shared with terminal (`result_model_schema_keys`).
    static let schemaKeys: [String] = [
        "raven_id", "display_name", "aliases", "profile_digest", "source_set",
        "verification_state", "introductions", "conflict_count", "sequence", "expires_at",
    ]
}

struct LocalDiscoveryContact: Equatable {
    var ravenId: String
    var pubHex: String
    var petname: String
    var publicTag: String
    var displayName: String
    var pinned: Bool
    var directlyVerified: Bool
}

struct SignedAliasClaim: Equatable {
    var alias: String
    var identityAddress: String
    var sequence: UInt64
    var expiresAt: UInt64
    var ed25519PubHex: String
}

/// Thin DiscoveryResolver-equivalent for iOS (same result model as ash / raven-core).
final class DiscoveryResolver {
    var contacts: [LocalDiscoveryContact] = []
    var aliasClaims: [SignedAliasClaim] = []
    var blockedPubHex: Set<String> = []
    /// When true (serverless), LegacyServer lane is never consulted.
    var serverless: Bool = true
    var publicProfileIndexEnabled: Bool = false
    var nowMs: UInt64 = 0

    static func normalizeAlias(_ raw: String) -> String? {
        let s = raw.trimmingCharacters(in: .whitespacesAndNewlines)
            .trimmingCharacters(in: CharacterSet(charactersIn: "@"))
            .lowercased()
        guard !s.isEmpty, s.count <= 64 else { return nil }
        let allowed = CharacterSet(charactersIn: "abcdefghijklmnopqrstuvwxyz0123456789_-")
        guard s.unicodeScalars.allSatisfy({ allowed.contains($0) }) else { return nil }
        return s
    }

    func search(query: String, scope: DiscoveryScope) -> [DiscoveryResult] {
        // Serverless discovery must never call FastAPI.
        if serverless {
            // no-op: LegacyServerProvider off
        }
        var out: [DiscoveryResult] = []
        let q = query.trimmingCharacters(in: .whitespacesAndNewlines)
        let runLocal = scope == .local || scope == .all || scope == .myNetwork
        let runId = scope == .exactId || scope == .all || scope == .publicScope
        let runAlias = scope == .exactAlias || scope == .all || scope == .publicScope
        let runNearby = scope == .nearby || scope == .all

        if runLocal {
            out.append(contentsOf: searchLocal(q))
        }
        if runId && q.hasPrefix("rvn1") {
            out.append(contentsOf: searchExactId(q))
        }
        if runAlias {
            out.append(contentsOf: searchExactAlias(q))
        }
        if runNearby {
            // Nearby results are injected by BLE confirm path; empty by default.
        }
        // PublicProfileIndex OFF in V1.
        _ = publicProfileIndexEnabled
        return merge(out)
    }

    private func searchLocal(_ query: String) -> [DiscoveryResult] {
        let q = query.trimmingCharacters(in: CharacterSet(charactersIn: "@")).lowercased()
        var out: [DiscoveryResult] = []
        for c in contacts {
            if blockedPubHex.contains(c.pubHex.lowercased()) {
                out.append(DiscoveryResult(
                    ravenId: c.ravenId, displayName: "", aliases: [], profileDigest: "",
                    sourceSet: [.localContacts], verificationState: .blocked,
                    introductions: [], conflictCount: 0, sequence: 0, expiresAt: 0
                ))
                continue
            }
            let match = c.ravenId.caseInsensitiveCompare(query) == .orderedSame
                || c.petname.lowercased().contains(q)
                || c.publicTag.lowercased() == q
                || c.displayName.lowercased().contains(q)
            guard match, !q.isEmpty || query.hasPrefix("rvn1") else { continue }
            out.append(DiscoveryResult(
                ravenId: c.ravenId,
                displayName: c.petname.isEmpty ? c.displayName : c.petname,
                aliases: c.publicTag.isEmpty ? [] : [c.publicTag],
                profileDigest: "",
                sourceSet: [.localContacts],
                verificationState: (c.directlyVerified || c.pinned) ? .directlyVerified : .trustedContact,
                introductions: [],
                conflictCount: 0,
                sequence: 0,
                expiresAt: UInt64.max
            ))
        }
        return out
    }

    private func searchExactId(_ query: String) -> [DiscoveryResult] {
        if let c = contacts.first(where: { $0.ravenId == query }) {
            if blockedPubHex.contains(c.pubHex.lowercased()) {
                return [DiscoveryResult(
                    ravenId: query, displayName: "", aliases: [], profileDigest: "",
                    sourceSet: [.exactRavenId], verificationState: .blocked,
                    introductions: [], conflictCount: 0, sequence: 0, expiresAt: 0
                )]
            }
        }
        return [DiscoveryResult(
            ravenId: query, displayName: "", aliases: [], profileDigest: "",
            sourceSet: [.exactRavenId], verificationState: .publicSignedProfile,
            introductions: [], conflictCount: 0, sequence: 0, expiresAt: 0
        )]
    }

    private func searchExactAlias(_ query: String) -> [DiscoveryResult] {
        guard let want = Self.normalizeAlias(query) else { return [] }
        let live = aliasClaims.filter {
            Self.normalizeAlias($0.alias) == want && $0.expiresAt >= nowMs
        }
        guard !live.isEmpty else { return [] }
        let conflict = live.count > 1 ? UInt32(live.count) : 0
        return live.map { claim in
            DiscoveryResult(
                ravenId: claim.identityAddress,
                displayName: "",
                aliases: [want],
                profileDigest: "",
                sourceSet: [.aliasDht],
                verificationState: conflict > 0 ? .aliasConflict : .publicSignedProfile,
                introductions: [],
                conflictCount: conflict,
                sequence: claim.sequence,
                expiresAt: claim.expiresAt
            )
        }
    }

    private func merge(_ results: [DiscoveryResult]) -> [DiscoveryResult] {
        var byId: [String: DiscoveryResult] = [:]
        for r in results {
            if r.verificationState == .aliasConflict {
                byId[r.ravenId] = r
                continue
            }
            if var e = byId[r.ravenId] {
                for s in r.sourceSet where !e.sourceSet.contains(s) {
                    e.sourceSet.append(s)
                }
                if e.displayName.isEmpty { e.displayName = r.displayName }
                for a in r.aliases where !e.aliases.contains(a) { e.aliases.append(a) }
                e.conflictCount = max(e.conflictCount, r.conflictCount)
                byId[r.ravenId] = e
            } else {
                byId[r.ravenId] = r
            }
        }
        return byId.values.sorted { $0.ravenId < $1.ravenId }
    }
}

File: node/crates/raven-core/src/discovery_resolver.rs
//! Multi-lane DiscoveryResolver — search without a central Raven DB.
//!
//! Finding a name ≠ verifying a person. `@alias` is NOT identity; `rvn1…` is.
//! Spec: `docs/RAVEN_DISCOVERY_V1.md`.

use crate::alias_record::{normalize_alias, AliasClaimStore, AliasRecord};
use crate::chat_history::BlockList;
use crate::introduction::IntroductionInbox;
use crate::nearby::NearbyRegistry;
use crate::profile_record::{ProfileStore, RavenProfileRecordV1};
use serde::{Deserialize, Serialize};
use std::collections::{BTreeSet, HashMap};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum DiscoveryScope {
    Local,
    ExactId,
    ExactAlias,
    MyNetwork,
    Nearby,
    Public,
    All,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum VerificationState {
    DirectlyVerified,
    TrustedContact,
    Introduced,
    ScopedVerified,
    NearbyVerified,
    PublicSignedProfile,
    AliasConflict,
    ExpiredOrStale,
    Blocked,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum DiscoverySource {
    LocalContacts,
    ExactRavenId,
    AliasDht,
    NearbyBle,
    SocialIntroduction,
    PublicProfileIndex,
    /// Present only when serverless flag is OFF — never required for V1 serverless.
    LegacyServer,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct DiscoveryIntroduction {
    pub introducer_raven_id: String,
    pub subject_raven_id: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct DiscoveryResult {
    pub raven_id: String,
    pub display_name: String,
    pub aliases: Vec<String>,
    pub profile_digest: String,
    pub source_set: Vec<DiscoverySource>,
    pub verification_state: VerificationState,
    pub introductions: Vec<DiscoveryIntroduction>,
    pub conflict_count: u32,
    pub sequence: u64,
    pub expires_at: u64,
}

/// Local contact row for LocalContactsProvider (bound to raven_id, not alias).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct LocalContactRow {
    pub raven_id: String,
    pub pub_hex: String,
    pub petname: String,
    pub public_tag: String,
    pub display_name: String,
    pub pinned: bool,
    pub directly_verified: bool,
}

/// Shared discovery context (in-process / community peer stand-in).
pub struct DiscoveryContext {
    pub contacts: Vec<LocalContactRow>,
    pub aliases: AliasClaimStore,
    pub profiles: ProfileStore,
    pub intros: IntroductionInbox,
    pub nearby: NearbyRegistry,
    pub blocked: BlockList,
    /// When true, LegacyServerProvider is disabled (serverless V1).
    pub serverless: bool,
    /// PublicProfileIndexProvider is STUB/OFF for V1.
    pub public_profile_index_enabled: bool,
    pub now_ms: u64,
}

impl Default for DiscoveryContext {
    fn default() -> Self {
        Self {
            contacts: Vec::new(),
            aliases: AliasClaimStore::default(),
            profiles: ProfileStore::default(),
            intros: IntroductionInbox::default(),
            nearby: NearbyRegistry::default(),
            blocked: BlockList::default(),
            serverless: true,
            public_profile_index_enabled: false,
            now_ms: 0,
        }
    }
}

pub trait DiscoveryProvider {
    fn name(&self) -> DiscoverySource;
    fn search(&self, ctx: &DiscoveryContext, query: &str) -> Vec<DiscoveryResult>;
}

pub struct LocalContactsProvider;
pub struct ExactRavenIdProvider;
pub struct AliasDhtProvider;
pub struct NearbyBleProvider;
pub struct SocialIntroductionProvider;
/// STUB/OFF for V1 — exact search first per research.
pub struct PublicProfileIndexProvider;
/// Must NOT be required when serverless flag is on.
pub struct LegacyServerProvider;

impl DiscoveryProvider for LocalContactsProvider {
    fn name(&self) -> DiscoverySource {
        DiscoverySource::LocalContacts
    }
    fn search(&self, ctx: &DiscoveryContext, query: &str) -> Vec<DiscoveryResult> {
        let q = query.trim().trim_start_matches('@').to_lowercase();
        let mut out = Vec::new();
        for c in &ctx.contacts {
            if ctx.blocked.is_blocked(&c.pub_hex) {
                out.push(blocked_result(&c.raven_id));
                continue;
            }
            let pet = c.petname.to_lowercase();
            let tag = c.public_tag.to_lowercase();
            let disp = c.display_name.to_lowercase();
            let id_match = c.raven_id.eq_ignore_ascii_case(query.trim());
            let local_match = !q.is_empty()
                && (pet.contains(&q)
                    || tag == q
                    || disp.contains(&q)
                    || c.raven_id.to_lowercase().contains(&q));
            if id_match || local_match {
                let state = if c.directly_verified || c.pinned {
                    VerificationState::DirectlyVerified
                } else {
                    VerificationState::TrustedContact
                };
                out.push(DiscoveryResult {
                    raven_id: c.raven_id.clone(),
                    display_name: if !c.petname.is_empty() {
                        c.petname.clone()
                    } else if !c.display_name.is_empty() {
                        c.display_name.clone()
                    } else {
                        c.public_tag.clone()
                    },
                    aliases: if c.public_tag.is_empty() {
                        vec![]
                    } else {
                        vec![c.public_tag.clone()]
                    },
                    profile_digest: String::new(),
                    source_set: vec![DiscoverySource::LocalContacts],
                    verification_state: state,
                    introductions: vec![],
                    conflict_count: 0,
                    sequence: 0,
                    expires_at: u64::MAX,
                });
            }
        }
        out
    }
}

impl DiscoveryProvider for ExactRavenIdProvider {
    fn name(&self) -> DiscoverySource {
        DiscoverySource::ExactRavenId
    }
    fn search(&self, ctx: &DiscoveryContext, query: &str) -> Vec<DiscoveryResult> {
        let q = query.trim();
        if !q.starts_with("rvn1") {
            return vec![];
        }
        if let Some(c) = ctx.contacts.iter().find(|c| c.raven_id == q) {
            if ctx.blocked.is_blocked(&c.pub_hex) {
                return vec![blocked_result(q)];
            }
        }
        if let Some(prof) = ctx.profiles.get(q, ctx.now_ms) {
            return vec![profile_to_result(prof, DiscoverySource::ExactRavenId, 0)];
        }
        // Valid-looking ID with no profile: still return candidate (exact ID path).
        if q.len() > 10 {
            return vec![DiscoveryResult {
                raven_id: q.to_string(),
                display_name: String::new(),
                aliases: vec![],
                profile_digest: String::new(),
                source_set: vec![DiscoverySource::ExactRavenId],
                verification_state: VerificationState::PublicSignedProfile,
                introductions: vec![],
                conflict_count: 0,
                sequence: 0,
                expires_at: 0,
            }];
        }
        vec![]
    }
}

impl DiscoveryProvider for AliasDhtProvider {
    fn name(&self) -> DiscoverySource {
        DiscoverySource::AliasDht
    }
    fn search(&self, ctx: &DiscoveryContext, query: &str) -> Vec<DiscoveryResult> {
        let raw = query.trim();
        // Exact alias lane: @name or normalized exact token (not fuzzy).
        let is_alias_query = raw.starts_with('@')
            || (normalize_alias(raw).is_ok() && !raw.starts_with("rvn1"));
        if !is_alias_query {
            return vec![];
        }
        let Ok(claims) = ctx.aliases.lookup_exact(raw, ctx.now_ms) else {
            return vec![];
        };
        if claims.is_empty() {
            return vec![];
        }
        let conflict_count = if claims.len() > 1 {
            claims.len() as u32
        } else {
            0
        };
        let mut out = Vec::new();
        for claim in claims {
            let mut r = claim_to_result(&claim, conflict_count, ctx);
            if conflict_count > 0 {
                r.verification_state = VerificationState::AliasConflict;
            }
            out.push(r);
        }
        out
    }
}

impl DiscoveryProvider for NearbyBleProvider {
    fn name(&self) -> DiscoverySource {
        DiscoverySource::NearbyBle
    }
    fn search(&self, ctx: &DiscoveryContext, _query: &str) -> Vec<DiscoveryResult> {
        let mut out = Vec::new();
        for b in &ctx.nearby.confirmed {
            out.push(DiscoveryResult {
                raven_id: b.peer_raven_id.clone(),
                display_name: String::new(),
                aliases: vec![],
                profile_digest: String::new(),
                source_set: vec![DiscoverySource::NearbyBle],
                verification_state: VerificationState::NearbyVerified,
                introductions: vec![],
                conflict_count: 0,
                sequence: 0,
                expires_at: u64::MAX,
            });
        }
        out
    }
}

impl DiscoveryProvider for SocialIntroductionProvider {
    fn name(&self) -> DiscoverySource {
        DiscoverySource::SocialIntroduction
    }
    fn search(&self, ctx: &DiscoveryContext, query: &str) -> Vec<DiscoveryResult> {
        let intros = ctx.intros.for_subject_alias(query, ctx.now_ms);
        let mut by_subject: HashMap<String, DiscoveryResult> = HashMap::new();
        for i in intros {
            let entry = by_subject.entry(i.subject_raven_id.clone()).or_insert_with(|| {
                DiscoveryResult {
                    raven_id: i.subject_raven_id.clone(),
                    display_name: i.subject_display_name.clone(),
                    aliases: i.subject_aliases.clone(),
                    profile_digest: String::new(),
                    source_set: vec![DiscoverySource::SocialIntroduction],
                    verification_state: VerificationState::Introduced,
                    introductions: vec![],
                    conflict_count: 0,
                    sequence: 0,
                    expires_at: i.expires_at,
                }
            });
            entry.introductions.push(DiscoveryIntroduction {
                introducer_raven_id: i.introducer_raven_id.clone(),
                subject_raven_id: i.subject_raven_id.clone(),
            });
        }
        by_subject.into_values().collect()
    }
}

impl DiscoveryProvider for PublicProfileIndexProvider {
    fn name(&self) -> DiscoverySource {
        DiscoverySource::PublicProfileIndex
    }
    fn search(&self, ctx: &DiscoveryContext, _query: &str) -> Vec<DiscoveryResult> {
        // STUB/OFF for V1 — no fuzzy / trigram index.
        if !ctx.public_profile_index_enabled {
            return vec![];
        }
        vec![]
    }
}

impl DiscoveryProvider for LegacyServerProvider {
    fn name(&self) -> DiscoverySource {
        DiscoverySource::LegacyServer
    }
    fn search(&self, ctx: &DiscoveryContext, _query: &str) -> Vec<DiscoveryResult> {
        if ctx.serverless {
            return vec![];
        }
        // Intentionally empty stub — FastAPI search must not be required.
        vec![]
    }
}

fn blocked_result(raven_id: &str) -> DiscoveryResult {
    DiscoveryResult {
        raven_id: raven_id.to_string(),
        display_name: String::new(),
        aliases: vec![],
        profile_digest: String::new(),
        source_set: vec![DiscoverySource::LocalContacts],
        verification_state: VerificationState::Blocked,
        introductions: vec![],
        conflict_count: 0,
        sequence: 0,
        expires_at: 0,
    }
}

fn profile_to_result(
    prof: &RavenProfileRecordV1,
    source: DiscoverySource,
    conflict_count: u32,
) -> DiscoveryResult {
    let state = if conflict_count > 0 {
        VerificationState::AliasConflict
    } else {
        VerificationState::PublicSignedProfile
    };
    DiscoveryResult {
        raven_id: prof.raven_id.clone(),
        display_name: prof.display_name.clone(),
        aliases: prof.public_aliases.clone(),
        profile_digest: hex::encode(prof.profile_digest()),
        source_set: vec![source],
        verification_state: state,
        introductions: vec![],
        conflict_count,
        sequence: prof.sequence,
        expires_at: prof.expires_at,
    }
}

fn claim_to_result(claim: &AliasRecord, conflict_count: u32, ctx: &DiscoveryContext) -> DiscoveryResult {
    let mut r = if let Some(prof) = ctx.profiles.get(&claim.identity_address, ctx.now_ms) {
        profile_to_result(prof, DiscoverySource::AliasDht, conflict_count)
    } else {
        DiscoveryResult {
            raven_id: claim.identity_address.clone(),
            display_name: String::new(),
            aliases: vec![claim.alias.clone()],
            profile_digest: String::new(),
            source_set: vec![DiscoverySource::AliasDht],
            verification_state: if conflict_count > 0 {
                VerificationState::AliasConflict
            } else {
                VerificationState::PublicSignedProfile
            },
            introductions: vec![],
            conflict_count,
            sequence: claim.sequence,
            expires_at: claim.expires_at,
        }
    };
    if !r.aliases.contains(&claim.alias) {
        r.aliases.push(claim.alias.clone());
    }
    r.sequence = claim.sequence;
    r.expires_at = claim.expires_at;
    if !r.source_set.contains(&DiscoverySource::AliasDht) {
        r.source_set.push(DiscoverySource::AliasDht);
    }
    // Local trusted contact upgrades verification (unless conflict/blocked).
    if let Some(c) = ctx
        .contacts
        .iter()
        .find(|c| c.raven_id == claim.identity_address)
    {
        if ctx.blocked.is_blocked(&c.pub_hex) {
            r.verification_state = VerificationState::Blocked;
        } else if conflict_count == 0 {
            r.verification_state = if c.directly_verified || c.pinned {
                VerificationState::DirectlyVerified
            } else {
                VerificationState::TrustedContact
            };
            if !r.source_set.contains(&DiscoverySource::LocalContacts) {
                r.source_set.push(DiscoverySource::LocalContacts);
            }
        }
    }
    r
}

/// Canonical multi-lane resolver.
pub struct DiscoveryResolver {
    providers: Vec<Box<dyn DiscoveryProvider + Send + Sync>>,
}

impl Default for DiscoveryResolver {
    fn default() -> Self {
        Self::v1()
    }
}

impl DiscoveryResolver {
    pub fn v1() -> Self {
        Self {
            providers: vec![
                Box::new(LocalContactsProvider),
                Box::new(ExactRavenIdProvider),
                Box::new(AliasDhtProvider),
                Box::new(NearbyBleProvider),
                Box::new(SocialIntroductionProvider),
                Box::new(PublicProfileIndexProvider),
                Box::new(LegacyServerProvider),
            ],
        }
    }

    pub fn search(&self, query: &str, scope: DiscoveryScope, ctx: &DiscoveryContext) -> Vec<DiscoveryResult> {
        let active: BTreeSet<DiscoverySource> = match scope {
            DiscoveryScope::Local => [DiscoverySource::LocalContacts].into_iter().collect(),
            DiscoveryScope::ExactId => [DiscoverySource::ExactRavenId].into_iter().collect(),
            DiscoveryScope::ExactAlias => [DiscoverySource::AliasDht].into_iter().collect(),
            DiscoveryScope::MyNetwork => [
                DiscoverySource::LocalContacts,
                DiscoverySource::SocialIntroduction,
            ]
            .into_iter()
            .collect(),
            DiscoveryScope::Nearby => [DiscoverySource::NearbyBle].into_iter().collect(),
            DiscoveryScope::Public => [
                DiscoverySource::ExactRavenId,
                DiscoverySource::AliasDht,
                DiscoverySource::PublicProfileIndex,
            ]
            .into_iter()
            .collect(),
            DiscoveryScope::All => self.providers.iter().map(|p| p.name()).collect(),
        };

        let mut merged: HashMap<String, DiscoveryResult> = HashMap::new();
        let mut conflict_bucket: HashMap<String, Vec<DiscoveryResult>> = HashMap::new();

        for p in &self.providers {
            if !active.contains(&p.name()) {
                continue;
            }
            // PublicProfileIndex stays OFF unless explicitly enabled.
            if p.name() == DiscoverySource::PublicProfileIndex && !ctx.public_profile_index_enabled
            {
                continue;
            }
            if p.name() == DiscoverySource::LegacyServer && ctx.serverless {
                continue;
            }
            for r in p.search(ctx, query) {
                if r.verification_state == VerificationState::AliasConflict {
                    conflict_bucket
                        .entry(query.trim().trim_start_matches('@').to_lowercase())
                        .or_default()
                        .push(r);
                    continue;
                }
                merge_result(&mut merged, r);
            }
        }

        // Surface all conflict candidates (never silent pick).
        let mut out: Vec<DiscoveryResult> = merged.into_values().collect();
        for (_alias, claims) in conflict_bucket {
            for r in claims {
                // Keep distinct raven_ids even under conflict.
                if let Some(existing) = out.iter_mut().find(|e| e.raven_id == r.raven_id) {
                    existing.verification_state = VerificationState::AliasConflict;
                    existing.conflict_count = existing.conflict_count.max(r.conflict_count);
                    for s in r.source_set {
                        if !existing.source_set.contains(&s) {
                            existing.source_set.push(s);
                        }
                    }
                } else {
                    out.push(r);
                }
            }
        }

        out.sort_by(|a, b| {
            rank(a.verification_state)
                .cmp(&rank(b.verification_state))
                .then(a.raven_id.cmp(&b.raven_id))
        });
        out
    }
}

fn merge_result(merged: &mut HashMap<String, DiscoveryResult>, r: DiscoveryResult) {
    merged
        .entry(r.raven_id.clone())
        .and_modify(|e| {
            for s in &r.source_set {
                if !e.source_set.contains(s) {
                    e.source_set.push(*s);
                }
            }
            if e.display_name.is_empty() && !r.display_name.is_empty() {
                e.display_name = r.display_name.clone();
            }
            for a in &r.aliases {
                if !e.aliases.contains(a) {
                    e.aliases.push(a.clone());
                }
            }
            if e.profile_digest.is_empty() {
                e.profile_digest = r.profile_digest.clone();
            }
            if rank(r.verification_state) < rank(e.verification_state) {
                e.verification_state = r.verification_state;
            }
            e.conflict_count = e.conflict_count.max(r.conflict_count);
            e.introductions.extend(r.introductions.iter().cloned());
            e.sequence = e.sequence.max(r.sequence);
            if r.expires_at > 0 {
                e.expires_at = if e.expires_at == 0 {
                    r.expires_at
                } else {
                    e.expires_at.min(r.expires_at)
                };
            }
        })
        .or_insert(r);
}

fn rank(s: VerificationState) -> u8 {
    match s {
        VerificationState::Blocked => 0,
        VerificationState::DirectlyVerified => 1,
        VerificationState::TrustedContact => 2,
        VerificationState::ScopedVerified => 3,
        VerificationState::Introduced => 4,
        VerificationState::NearbyVerified => 5,
        VerificationState::PublicSignedProfile => 6,
        VerificationState::AliasConflict => 7,
        VerificationState::ExpiredOrStale => 8,
    }
}

/// Stable JSON shape for terminal/mobile parity (acceptance #19).
pub fn result_model_schema_keys() -> &'static [&'static str] {
    &[
        "raven_id",
        "display_name",
        "aliases",
        "profile_digest",
        "source_set",
        "verification_state",
        "introductions",
        "conflict_count",
        "sequence",
        "expires_at",
    ]
}
